### DISCUSSION TITLE
Print a partial schema/TypeDefinitionRegistry to SDL

### DISCUSSION THREAD

**Discussion by seanf** [graphql-java/graphql-java#2768] [Q&A]
I have some code (`com.netflix.graphql.dgs.pagination.DgsPaginationTypeDefinitionRegistry.registry()`) which takes an original `TypeDefinitionRegistry` and generates an extra `TypeDefinitionRegistry` with some derived types. The derived types use some of the original types in their definitions. I would like to write out the extra types to a separate schema file (SDL), without the contents of the original schema.

If I call `SchemaGenerator.makeExecutableSchema()` on the extra types, I get errors about the missing original types, so I have had to create a `TypeDefinitionRegistry` which includes both the original types and the extra types. But of course the resulting SDL via `SchemaPrinter` includes both original and extra types. 

Is there any way I can generate SDL from the extra types only?

**Comment by bbakerman** [comment]
> If I call SchemaGenerator.makeExecutableSchema() on the extra types, I get errors about the missing original types

This is mandatory right - a schema is not valid without all its types. 

> Is there any way I can generate SDL from the extra types only?


We have done something like this at our work.  I am not sure if it will work for you but here goes.

We needed to print out types that came from team A and remove types that came from a central schema place (common directives and so on)

We did this by looking at the `graphql.language.SourceLocation` that is inside the AST definition object per type.

For example the AST elements that define a `graphql.schema.GraphQLObjectType` are held in a `graphql.language.ObjectTypeDefinition` and that has a SourceLocation

We then used out internal naming where by the original source files structure was used to allot ownership.

When printing there is a `graphql.schema.idl.SchemaPrinter.Options#getIncludeSchemaElement` predicate that can "filter" what gets printed

The other alternative is to create a map of  schema element -> source type registry during the SchemaGenerator build process.

a callback `graphql.schema.idl.SchemaGeneratorPostProcessing` could be used say but you need to have captured all the type registries yourself - they are not state passed back.

You could then iterate all the types in a  schema, look them up by name in a given type registry and then map the association of runtime type -> type registry source place.

The use the `graphql.schema.idl.SchemaPrinter.Options#getIncludeSchemaElement` predicate to decide which ones you print say.
